### PR TITLE
API: `optimize.differential_evolution`: transition to Generator (SPEC 7)

### PR DESCRIPTION
**Pull request by andyfaff**
Transitions `optimize.differential_evolution` according to [SPEC007](https://scientific-python.org/specs/spec-0007/#code).

@mdhaber, @tupui.

### CODE REVIEW

File: scipy/_lib/_util.py
Comment (andyfaff) [inline scipy/_lib/_util.py:436]: I tried changing the use in `differential_evolution` from this version to the one in `stats._qmc`. However, this results in circular imports. The version in `stats._qmc` would preferably be changed to this `check_random_state_spec007` version.

**Review by tupui — APPROVED** [review]
Big yes for me. I have eggs in the basket so I will let the crowd decide. Approving if that counts in any ways.

File: scipy/optimize/tests/test__differential_evolution.py
Comment (j-bowhay) [inline scipy/optimize/tests/test__differential_evolution.py:526]: Should we test every function we apply this to or have a central test on a dummy function?

Comment (andyfaff) [reply scipy/optimize/tests/test__differential_evolution.py:526]: We should have a good central test that exercises the decorator. This does to a certain extent. I don't think we necessarily need to do it for every function. However, there may be remnant code that wasn't written for `Generator`, in which case some extra testing wouldn't go amiss.

**Comment by andyfaff** [comment]
The biggest effect of PRs of this type will be the default RNG will be a `Generator`.

Previously if one supplied an `int` then the seeded RNG would be `RandomState`, now it'll be `Generator`.
Previously if one didn't specify seed then the RNG would be `RandomState`.

This is going to change random number generation, and code that relied on a certain seed for reproducibility will be different across versions.

File: scipy/_lib/_util.py
Comment (mdhaber) [inline scipy/_lib/_util.py:249]: I checked that this is copied almost veratim, but changes are intentional. Consider submitting them upstream.

File: scipy/optimize/_differentialevolution.py
Comment (mdhaber) [inline scipy/optimize/_differentialevolution.py:24]:
```suggestion
@_transition_to_rng("seed", position_num=9)
```
Right? Looks like `seed`/`rng` was/is at position 9 (0-indexed).

File: scipy/_lib/_util.py
Comment (mdhaber) [inline scipy/_lib/_util.py:410]: I don't think we need this yet. According to the SPEC:
- For the first two cycles, nothing changes except `rng` and `seed` are both accepted.
- Then we add an `end_version`, so for another two cycles, a deprecation warning is emitted that `seed` will no longer be accepted.
- The only time behavior changes is at the end of the four cycles - we remove the old `check_random_state` or whatever it is called and replace it with just `np.random.default_rng`.

**Comment by mdhaber** [comment]
Thanks @andyfaff for doing this!

> This is going to change random number generation, and code that relied on a certain seed for reproducibility will be different across versions.

The intent of the SPEC is that this only happens after warnings expire. In the meantime, arguments passed positionally or via the old keyword argument are processed as they always were, whatever that is. (Arguments passed using the new keyword are validated and standardarized by the decorator according to the new behavior, which is simply to call `rng = np.random.default_rng(rng)`.) 

After the warnings have been emitted for two cycles (according to SciPy's policy), then we can remove the decorator and any old argument validation/standardization code except for `rng = np.random.default_rng(rng)`.

File: scipy/optimize/_differentialevolution.py
Comment (andyfaff) [reply scipy/optimize/_differentialevolution.py:24]: Good catch. How on earth did the tests pass then.

File: scipy/_lib/_util.py
Comment (andyfaff) [reply scipy/_lib/_util.py:410]: Ok. Understood. I can remove this function, and go back to using the original.

**Comment by andyfaff** [comment]
@mdhaber what the SPEC doesn't provide guidelines on is what should happen to the docstring of `differential_evolution`. For the initial state should there still be documentation on what happens if `seed` (`random_state` in the SPEC example) is specified? Or should it just specify the `rng` behaviour? Should it mention the SPEC?

**Comment by andyfaff** [comment]
Ok, I think I got a useful docstring now.

**Comment by mdhaber** [comment]
> @mdhaber what the SPEC doesn't provide guidelines on is what should happen to the docstring

It was actually an explicit decision to leave that up to projects due to differences in opinion over how (/whether) the transition should be documented explicitly. One side is that only the new behavior needs to be documented, since old code will continue to work and new uses should follow the new pattern. The other is that we need to document all supported behaviors of the function - both as a matter of principle and to support those who are referring to the documentation while working with existing code that uses the old behavior.

Regardless of the level of detail you'd prefer to go into about the supported old behavior, I'd suggest leading with the new behavior to encourage its use. These things are true whether the argument is passed by position or by keyword:
```python3
    rng : `numpy.random.Generator`, optional
        Pseudorandom number generator state. When `rng` is None, a new
        `numpy.random.Generator` is created using entropy from the
        operating system.
```

Then I think it's OK to document all the details if you prefer. In this interim period:
- If `rng` is passed by keyword, types other than `numpy.random.Generator` are passed to `numpy.random.default_rng` to instantiate a `Generator`.
- If the argument is passed by position or `seed` is passed by keyword, the behavior is whatever the old behavior was (to maintain backward compatibility).

Thanks for pioneering this transition. I know it's messy!

File: scipy/optimize/_differentialevolution.py
Comment (mdhaber) [inline scipy/optimize/_differentialevolution.py:613]: Looks like this is missing the heading of the `rng` argument.

**Comment by andyfaff** [comment]
How about:

```
    rng : {None, int, `numpy.random.Generator`}, optional
        
        ..versionchanged:: 1.15.0
            As part of the `SPEC-007 <https://scientific-python.org/specs/spec-0007/>`_
            transition from use of `numpy.random.RandomState` to
            `numpy.random.Generator` this keyword was changed from `seed` to `rng`.
            For an interim period both keywords will continue to work (only specify
            one of them). After the interim period using the `seed` keyword will emit
            warnings. The behavior of the `seed` and `rng` keywords is outlined below.

        If `rng` is passed by keyword, types other than `numpy.random.Generator` are
        passed to `numpy.random.default_rng` to instantiate a `Generator`.
        If `rng` is already a `Generator` instance, then the provided instance is
        used.
        
        If this argument is passed by position or `seed` is passed by keyword, the
        behavior is:
        
        - If `seed` is None (or `np.random`), the `numpy.random.RandomState`
        singleton is used.
        - If `seed` is an int, a new `RandomState` instance is used,
        seeded with `seed`.
        - If `seed` is already a `Generator` or `RandomState` instance then
        that instance is used.
        
        Specify `seed`/`rng` for repeatable minimizations.
```

**Comment by mdhaber** [comment]
I think that's OK. 

Just some rendering issues, especially with the admonition:
image

We never saw this rendered, so I think we need to do something about the single backticks around plain `Generator` in the SPEC 7 example. Maybe there's a way to get it to link as intended without specifying the full name (`numpy.random.Generator`) every time, but intersphinx always confuses me. There are a lot more uses of `Generator`, `RandomState`, and `np.random` here that are probably supposed to render as links.

@stefanv Just thought I'd ping you so you can see how one of these SPEC 7 transitions to `rng` is playing out.

**Comment by andyfaff** [comment]
I've tried to improve by putting more backticks around e.g. ``Generator``. We don't need multiple links everywhere it's mentioned, and using something like :class:`np.random.Generator` makes it harder to look at when looking at the plain text version.

**Comment by andyfaff** [comment]
I don't know why the versionchanged isn't displayed properly.

**Comment by andyfaff** [comment]
It's rendering properly now.

**Review by mdhaber — APPROVED** [review]
I think this is ready. If you're happy with the final tweaks @andyfaff, I'll go ahead and merge. SciPy already endorsed SPEC 7.

**Comment by lucascolley** [comment]
> I'll go ahead and merge. SciPy already endorsed SPEC 7.

Fine to merge, but it wasn't decided in the endorsement thread that we were also happy to implement the SPEC everywhere in SciPy (they are two separate things). The relevant news from the forum is that [Andrew's post for this API change](https://discuss.scientific-python.org/t/implementing-spec007-for-differential-evolution/1514) received no pushback.

**Comment by andyfaff** [comment]
@mdhaber, I'm happy with the changes you made.

Once this gets merged I'm intending to start working on similar changes. First the other global optimisers, then `stats.qmc`.

I think it would be good to do all the stats distributions as well, but that may be a big task.

**Comment by mdhaber** [comment]
Thanks @andyfaff!

File: scipy/optimize/tests/test__differential_evolution.py
Comment (j-bowhay) [inline scipy/optimize/tests/test__differential_evolution.py:549]: I think these need to be in separate context managers as the test will still pass if only one these raises an error

Comment (mdhaber) [reply scipy/optimize/tests/test__differential_evolution.py:549]: Yup. Let's fix that in the next PR like this.

**Pull request merged**
